code/controllers/ContentController.php

```diff
@@ -87,42 +88,58 @@ public function Page($link) {
 
 	public function init() {
 		parent::init();
-		
+
+		$recordExists = ($this->dataRecord && $this->dataRecord->exists());
+
 		// If we've accessed the homepage as /home/, then we should redirect to /.
-		if($this->dataRecord && $this->dataRecord instanceof SiteTree
-			 	&& RootURLController::should_be_on_root($this->dataRecord) && (!isset($this->urlParams['Action']) || !$this->urlParams['Action'] ) 
-				&& !$_POST && !$_FILES && !$this->redirectedTo() ) {
+		if(
+			$recordExists
+			&& RootURLController::should_be_on_root($this->dataRecord)
+			&& !$this->request->param('Action')
+			&& !$this->request->postVars()
+			&& !$this->redirectedTo()
+		) {
 			$getVars = $_GET;
 			unset($getVars['url']);
-			if($getVars) $url = "?" . http_build_query($getVars);
-			else $url = "";
+			if($getVars) {
+				$url = "?" . http_build_query($getVars);
+			} else {
+				$url = "";
+			}
 			$this->redirect($url, 301);
 			return;
 		}
-		
-		if($this->dataRecord) $this->dataRecord->extend('contentcontrollerInit', $this);
-		else singleton('SiteTree')->extend('contentcontrollerInit', $this);
 
-		if($this->redirectedTo()) return;
+		// Run extension hooks
+		if($recordExists) {
+			$this->dataRecord->extend('contentcontrollerInit', $this);
+		} else {
+			singleton('SiteTree')->extend('contentcontrollerInit', $this);
```

Comment:
See my note above on the Security controller. Do you think extending `SiteTree` over `Page` is a good idea? Since we are in the CMS module it's probably ok to rely on there being a Page class.